services/horizon/internal/render/sse/main.go

```diff
@@ -132,6 +133,55 @@ func getJSON(val interface{}) string {
 	return string(js)
 }
 
+// Support notifying open SSE requests via channel based on related changes done
+// in the History database. Channel also includes reference counting for safe removal.
+type PubsubChannel struct {
+	channel        chan struct{}
+	referenceCount int
+}
+
+var pubsub_channels = make(map[string]PubsubChannel)
+
+// SSE connection can subscribe to a topic which is ususllay an id (account, ledger, tx)
+// Once a change in database happens, Publish is used by ingestor so channel is notified.
+func Subscribe(topic string) chan struct{} {
+	if _, ok := pubsub_channels[topic]; !ok {
+		pubsub_channels[topic] = PubsubChannel{make(chan struct{}), 1}
+	} else {
+		pubsub_channels[topic] = PubsubChannel{pubsub_channels[topic].channel, pubsub_channels[topic].referenceCount + 1}
+	}
+	log.Infof("Subscribe to topic: %s", topic)
+	return pubsub_channels[topic].channel
+}
+
+// Unsubscribe to a topic, fir example when SSE connectio is closed. Reference counting is used in
+// order to decide if channel should stay alive or closed.
+func Unsubscribe(topic string) {
+	if pubsubChannel, ok := pubsub_channels[topic]; ok {
+		if pubsubChannel.referenceCount == 1 {
+			close(pubsubChannel.channel)
+			delete(pubsub_channels, topic)
```

Comment:
what happens if this condition is reached twice in parallel?

Reply:
Changed implementation